// monaco-react/src/Editor.tsx
import * as React from 'react';
import * as monaco from 'monaco-editor';
import { editor } from 'monaco-editor';
import { IStaticConfig } from './IStaticConfig';
import { environment } from './environment';
import { IEditorProps } from './IEditorProps';

import './Editor.less';

export class Editor extends React.Component<IEditorProps, any> {

  static config(opts: IStaticConfig) {
    environment.baseUrl = opts.baseUrl;
  }

  monaco: editor.IStandaloneCodeEditor;
  container: any;
  setContainerRef = (ref: any) => this.container = ref;

  render() {
    return <div className="monaco-container" ref={this.setContainerRef}></div>;
  }

  private createOptions = (props: IEditorProps) => {
    const options: IEditorProps = Object.assign({
      value: '',
      language: 'javascript',
      automaticLayout: true,
      readOnly: false,
      minimap: { enabled: false },
      folding: true,
      renderLineHighlight: 'none'
    }, props);
    return options;
  };

  private init = () => {
    if (!this.container) return;
    const options = this.createOptions(this.props);
    this.monaco = editor.create(this.container, options);
    this.monaco.getModel().updateOptions({ tabSize: 2 });
    this.monaco.layout();
  };

  private bindEvents = () => {
    if (!this.monaco) return;
    const { onChange } = this.props;
    if (!onChange) return;
    this.monaco.onDidChangeModelContent(event => {
      const value = this.monaco.getValue();
      onChange(value, this.monaco, this);
    });
  };

  componentDidMount() {
    this.init();
    this.bindEvents();
    const { onReady } = this.props;
    if (onReady) onReady(this.monaco, this);
  }

  componentWillUnmount() {
    if (this.monaco) this.monaco.dispose();
  }

  getSelectedText = () => {
    if (this.monaco) return;
    const textModel = this.monaco.getModel()
    return textModel.getValueInRange(this.monaco.getSelection());
  }

}

export default Editor;
